### PR TITLE
Upgrade support from v0.5.0 to v0.6.0 

### PR DESCRIPTION
**Pull request by shuo-wu**
https://github.com/longhorn/longhorn/issues/622

### CODE REVIEW

File: controller/engine_controller.go [hunk 1]
```diff
@@ -266,26 +266,44 @@ func (ec *EngineController) syncEngine(key string) (err error) {
 		}
 	}()
 
+	var isCLIAPIVersionOne bool
+	if engine.Status.CurrentImage != "" {
```
Comment (yasker) [inline controller/engine_controller.go]: Why do we need to check the `CurrentImage`? The CurrentImage should be set if the volume is running.

Comment (shuo-wu) [reply controller/engine_controller.go]: I want to make sure that the empty string won't be used in the function. Then I can guarantee that the related fields are cleaned up for the incompatible detached volume.

Comment (yasker) [reply controller/engine_controller.go]: I don't follow. Is `CurrentImage==""` valid or not? In which situation we will fall back to use `Spec.EngineImage`?

Comment (shuo-wu) [reply controller/engine_controller.go]: Sometimes `CurrentImage` is cleaned up but `Endpoint` and `ReplicaModeMap` retain there. I will double-check the reason.

Comment (shuo-wu) [reply controller/engine_controller.go]: Cannot reproduce it... It should be fine to just use `CurrentImage` only without `Spec.EngineImage` as fallback. But I am afraid that there may be unexpected corner cases that will cause the above bug.

File: controller/engine_controller.go [hunk 2]
```diff
@@ -266,26 +266,39 @@ func (ec *EngineController) syncEngine(key string) (err error) {
...
-			}
-		} else if engine.Status.ReplicaModeMap != nil {
-			if err := ec.ReconcileEngineState(engine); err != nil {
-				return err
+	// For incompatible engine, skip starting engine monitor and clean up fields when the engine is not running
+	if isCLIAPIVersionOne {
+		if engine.Status.CurrentState != types.InstanceStateRunning {
+			engine.Status.Endpoint = ""
+			engine.Status.ReplicaModeMap = nil
+		}
```
Comment (yasker) [inline controller/engine_controller.go]: Add `return` to avoid changing the main flow of code.

File: datastore/longhorn.go [hunk 3]
```diff
@@ -1427,3 +1427,20 @@ func (s *DataStore) GetInstanceManagerByInstance(obj interface{}) (*longhorn.Ins
 	engineImageName := types.GetEngineImageChecksumName(engineImage)
 	return s.GetInstanceManagerBySelector(nodeID, engineImageName, imType)
 }
+
+func (s *DataStore) IsEngineImageCLIAPIVersionOne(imageName string) (bool, error) {
+	if imageName == "" {
```
Comment (yasker) [inline datastore/longhorn.go]: Why isn't an error?

File: controller/engine_controller.go [hunk 4]
```diff
@@ -426,6 +431,61 @@ func (ec *EngineController) DeleteInstance(obj interface{}) error {
...
 }
 
+func (ec *EngineController) deleteInstanceWithCLIAPIVersionOne(e *longhorn.Engine) (err error) {
+	isCLIAPIVersionOne, err := ec.ds.IsEngineImageCLIAPIVersionOne(e.Status.CurrentImage)
+	if err != nil {
+		return err
+	}
+
+	if isCLIAPIVersionOne {
+		pod, err := ec.podLister.Pods(ec.namespace).Get(e.Name)
```
Comment (yasker) [inline controller/engine_controller.go]: Don't need the informer. Use API directly for the old version - it shouldn't last long.

File: controller/controller_manager.go [hunk 5]
```diff
@@ -85,10 +85,10 @@ func StartControllers(stopCh chan struct{}, controllerID, serviceAccount, manage
 		persistentVolumeInformer, persistentVolumeClaimInformer,
 		kubeClient, namespace)
 	rc := NewReplicaController(ds, scheme,
-		replicaInformer, imInformer,
+		podInformer, replicaInformer, imInformer,
```
Comment (yasker) [inline controller/controller_manager.go]: Don't think we need the informer. Just use API directly.

File: controller/engine_controller.go (hunk 2 above)
Comment (yasker) [reply controller/engine_controller.go]:
```
if isCLIAPIVersionOne {
	...
         return nil
}
if engine.status.CurrentState == xxxx
```

File: controller/volume_controller_test.go [hunk 6]
```diff
@@ -847,6 +847,12 @@ func (s *TestSuite) runTestCases(c *C, testCases map[string]*VolumeTestCase) {
 		c.Assert(err, IsNil)
 		pIndexer.Add(p)
 
+		ei, err := lhClient.LonghornV1alpha1().EngineImages(TestNamespace).Create(newEngineImage(types.EngineImageStateReady))
```
Comment (yasker) [inline controller/volume_controller_test.go]: Why we need those?

Comment (shuo-wu) [reply controller/volume_controller_test.go]: `IsEngineImageCLIAPIVersionOne()` is used in volume controller and it needs the engine image to check the version

**Pull request merged**